From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: daskey <13271393+daskey@users.noreply.github.com>
Date: Fri, 26 Dec 2025 03:24:20 +0800
Subject: [PATCH] 8828: PowerPC: Add SPE floating point min/max instructions
 (efsmax, efsmin)

PowerPC: Add SPE floating point min/max instructions
 (efsmax, efsmin)

Refactor efsmax and efsmin for clarity and efficiency
---
 .../PowerPC/data/languages/Scalar_SPFP.sinc   | 31 +++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/Ghidra/Processors/PowerPC/data/languages/Scalar_SPFP.sinc b/Ghidra/Processors/PowerPC/data/languages/Scalar_SPFP.sinc
index 6f812bcc43..da8ba5dfb5 100644
--- a/Ghidra/Processors/PowerPC/data/languages/Scalar_SPFP.sinc
+++ b/Ghidra/Processors/PowerPC/data/languages/Scalar_SPFP.sinc
@@ -467,3 +467,34 @@ define pcodeop ConvertFloatingPointFromUnsignedFraction;
 {
   CRFD[2,1] = A:4 f< B:4;
 }
+
+# =================================================================
+# SPE Floating Point Single-Precision Min/Max
+
+# efsmax rT,rA,rB      010 1011 0000
+#define pcodeop FloatingPointMaximum;
+:efsmax D,A,B is OP=4 & D & A & B & XOP_0_10=0x2B0
+{
+   local inA:4 = A:4;
+   local inB:4 = B:4;
+   local cond:1 = inA f>= inB;
+   local result:4 = (zext(cond) * inA) + (zext(!cond) * inB);
+   # assign to lower word of D
+   D = ( D & 0xFFFFFFFF00000000 ) | zext( result );
+   setFPRF( result );
+   setSummaryFPSCR();
+}
+
+# efsmin rT,rA,rB      010 1011 0001
+#define pcodeop FloatingPointMinimum;
+:efsmin D,A,B is OP=4 & D & A & B & XOP_0_10=0x2B1
+{
+   local inA:4 = A:4;
+   local inB:4 = B:4;
+   local cond:1 = inA f<= inB;
+   local result:4 = (zext(cond) * inA) + (zext(!cond) * inB);
+   # assign to lower word of D
+   D = ( D & 0xFFFFFFFF00000000 ) | zext( result );
+   setFPRF( result );
+   setSummaryFPSCR();
+}
-- 
2.45.1

